beacon-chain/cache/attestation.go

```diff
@@ -0,0 +1,264 @@
+package cache
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v5/beacon-chain/operations/attestations/forkchoice"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/v5/crypto/bls"
+	ethpb "github.com/prysmaticlabs/prysm/v5/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v5/proto/prysm/v1alpha1/attestation"
+)
+
+type attGroup struct {
+	slot     primitives.Slot
+	local    ethpb.Att
+	external []ethpb.Att
+}
+
+type AttestationCache struct {
+	atts map[attestation.Id]*attGroup
+	sync.RWMutex
+	forkchoiceAtts *forkchoice.Attestations
+}
+
+func NewAttestationCache() *AttestationCache {
+	return &AttestationCache{
+		atts:           make(map[attestation.Id]*attGroup),
+		forkchoiceAtts: forkchoice.New(),
+	}
+}
+
+func (c *AttestationCache) Add(att ethpb.Att) error {
+	if att.IsNil() {
+		return nil
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	id, err := attestation.NewId(att, attestation.Data)
+	if err != nil {
+		return errors.Wrapf(err, "could not create attestation ID")
+	}
+
+	group := c.atts[id]
+	if group == nil {
+		group = &attGroup{
+			slot: att.GetData().Slot,
+		}
+		c.atts[id] = group
+	}
+
+	if att.IsAggregated() {
+		group.external = append(group.external, att.Clone())
```

Comment:
Perhaps document/comment that we are not checking if the attestation is fully covered because we are dropping those on gossip. But I think we may need to drop those when we handle the operations from reorged blocks.